tools/cache_testing.py

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python
+
+import redis
+import argparse
+from urllib.parse import urlparse
+import numpy as np
+
+'''
+Run Cache Testing.
+This tool performs cache testing for Dragonfly
+by calling the `incrby` function on a constrained set
+of items, as defined by the user. Additionally, it
+distributes the frequency of `incrby` calls for each
+item based on a Zipfian distribution (with alpha values
+between 0 and 1 being representative of real-life cache
+load scenarios)
+'''
+
+def rand_zipf(n, alpha, numSamples):
+    """
+    n: The upper bound of the values to generate a zipfian distribution over
+    (n = 30 would generate a distribution of given alpha from values 1 to 30)
+    alpha: The alpha parameter to be used while creating the Zipfian distribution
+    num_samples: The total number of samples to generate over the Zipfian distribution
+    returns a list of num_samples values of items in the range [1, n] that follow a
+    Zipfian distribution.
+    """
+
+    # Calculate Zeta values from 1 to n:
+    tmp = np.power( np.arange(1, n+1), -alpha )
+    zeta = np.r_[0.0, np.cumsum(tmp)]
+
+    # Store the translation map:
+    distMap = [x / zeta[-1] for x in zeta]
+
+    # Generate an array of uniform 0-1 pseudo-random values:
+    u = np.random.random(numSamples)
```

Comment:
I understand that you took the implementation from here:
https://stackoverflow.com/questions/31027739/python-custom-zipf-number-generator-performing-poorly

it's fine, but I am worried about the memory usage of the this program - we will need to generate traces of hundreds of millions/billions of records to cause evictions.  To solve this, we should convert rand_zipf to a generator, yielding 
batches of `np.random.random(numSamples)` each time and then we can generate a batch each time repeating the preprocessor step every time I call `rand_zipf`.  if you know what I am talking about please do the adjustments otherwise leave as is.

Reply:
I have fixed the `numSamples` bottleneck by rewriting the code to look something like this:

```bash
    # Calculate Zeta values from 1 to n:
    tmp = np.power( np.arange(1, n+1), -alpha )
    zeta = np.r_[0.0, np.cumsum(tmp)]

    # Store the translation map:
    distMap = [x / zeta[-1] for x in zeta]

    # Generate values form the distribution based on 0-1 pseudo-random values

    for i in range(numSamples):
        yield np.searchsorted(distMap, np.random.random())
```

However, it's worth nothing that there exists another significant bottleneck -- the distMap. It's likely going to also be a huge data structure (if we really want to start seeing cache misses, that is). I think it might be a bit more challenging to remove that bottleneck, but it may be possible (I still have to spend some more time thinking about how, if at all, it can be done). Do you think I should spend time on it? Because another workaround to forcing misses out of the cache is to simply limit the amount of memory it is allowed, too.

Reply:
I actually thought to leave numSamples batch as is but instead to have:
```py
while True:
   u = np.random.random(numSamples)
    v = np.searchsorted(distMap, u)
    samples = [t-1 for t in v]
    yield samples
```

and use pipeline length as numSamples.

Reply:
I would not worry about distMap right now. yes it's big but I think it's smaller than total number of writes we gonna send to the server.

Reply:
Ah, I understand. Shall fix accordingly!